### ISSUE TITLE
Our build script tests hang when run with Mono on Windows 

### ISSUE THREAD

**Issue by jnm2** [nunit/nunit#3222]
When I run the equivalent of test45 using Mono 5.20 on Windows just like we do on Linux, the test run hangs consistently. It looks similar to what we've seen intermittently in CI.

Steps to reproduce from a PowerShell prompt on Windows:

- `choco install mono -y` if you don't have it already

- `.\build.ps1 -t build`

- ```
  &"C:\Program Files\Mono\bin\mono.exe" `
    tools/NUnit.ConsoleRunner.3.9.0/tools/nunit3-console.exe `
    bin/Release/net45/nunit.framework.tests.dll `
    --result=test-results/net45/nunit.framework.tests.xml `
    --process=InProcess
  ```

The test run has never completed for me.

**Comment by jnm2** [comment]
By adding `--labels=Before`, I can see that it's hanging during ` NUnit.Framework.Attributes.TimeoutTests.TimeoutWithMessagePumpShouldAbort`. This is a test I added in https://github.com/nunit/nunit/pull/2023. I'm going to try to fix this right away.

**Comment by jnm2** [comment]
The test passes when it is the only test in the test run but not when grouped with all the others.

**Comment by jnm2** [comment]
If I remove the WaitMessage code from it so that it just `while (true) { }` like the other test next to it, it still deadlocks.

**Comment by jnm2** [comment]
Wild hypothesis, but it's starting to feel like Mono never throws ThreadAbortException again once a thread had called ResetAbort even if it's aborted again. If that's the case, you could never have more than one timed out test per thread or else the next timeout would hang.

**Comment by jnm2** [comment]
I went so far as to remove all the aborting logic I added in https://github.com/nunit/nunit/pull/2023 but running one extra normal timeout test still hangs when it should be timing out. Thread.CurrentThread.ThreadState is AbortRequested, but no ThreadAbortException gets raised.

**Comment by jnm2** [comment]
If my hypothesis is correct, that would explain the intermittency. To fail, you have to be on Mono and have more than one of the timeout tests get picked up by the same worker thread.

**Comment by CharliePoole** [comment]
Crazy. You could write a test for that, of course. @migueldeicaza does that sound possible to you? @jnm2 what version of mono?

**Comment by jnm2** [comment]
@CharliePoole I haven't verified that. I'm working on it as we speak. But I just had a thought. In https://github.com/nunit/nunit/issues/3221#issuecomment-487405960, a stray ThreadAbortException comes along out of nowhere. Here, none shows up. I wonder if somehow something is delaying it so that it only shows up when it's no longer relevant.

Unfortunately I'll be stopping for the evening in ten minutes due to another commitment. I don't mind the break though; I've been banging my head against the wall for six hours at this point. In light of that, everything I say should probably be taken with a grain of salt now that I think of it. :D

**Comment by jnm2** [comment]
@CharliePoole Sorry, forgot to add the `--version`:

```
Mono JIT compiler version 5.20.1 (Visual Studio built mono)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
        TLS:           __thread
        SIGSEGV:       normal
        Notification:  Thread + polling
        Architecture:  amd64
        Disabled:      none
        Misc:          softdebug
        Interpreter:   yes
        LLVM:          supported, not enabled.
        Suspend:       preemptive
        GC:            sgen (concurrent by default)
```

**Comment by jnm2** [comment]
If we are going to support Mono on any platform, I'd really like to be running our tests against Mono on Windows. This is what I think we should be testing on Windows:

- Our net35 DLL against .NET CLR v2, .NET CLR v4, and Mono (3 runs)
- Our net40 DLL against .NET CLR v4 and Mono (2 runs)
- Our net45 DLL against .NET CLR v4 and Mono (2 runs)
- Our netstandard1.4 DLL against .NET CLR v4, .NET Core v1.1, .NET Core v2.1, .NET Core v2.2, and Mono (5 runs)
- Our netstandard2.0 DLL against .NET CLR v4, .NET Core v2.1, .NET Core v2.2, and Mono (4 runs)

This is all just for our Windows build script. Linux and macOS should do the same but drop the .NET Framework runs of course.

Otherwise, we just won't know for sure how our DLLs behave on these currently untested combinations.

**Comment by jnm2** [comment]
Now that I'm coming back with fresh eyes, I'm noticing that a number of our timeout tests have this line:

```cs
[Platform(Exclude = "Mono", Reason = "Runner hangs at end when this is run")]
```

This was introduced ten years ago: https://github.com/nunit/nunit/commit/fc042531ab73cfebc0c1b71e1bb26ff2bc2483dc
It looks as though Mono has been consistently problematic since then. What I missed in #2023 was that I should have excluded Mono for my test as well, since it's just a fancier version of:

https://github.com/nunit/nunit/blob/9f2d8ea7c35518ee28f1255bc21aeca6d84057b3/src/NUnitFramework/tests/Attributes/TimeoutTests.cs#L65-L76

**Comment by jnm2** [comment]
I nailed down the actual bug in Mono and worked around it so that none of these tests need to be ignored on Mono anymore!

To bring closure to this issue, more details starting at https://github.com/nunit/nunit/pull/3095#issuecomment-491555544.

tl;dr: Under some circumstances, if you swallow a ThreadAbortException without calling ResetAbort, Mono does not throw ThreadAbortException when the control flow leaves the catch block. Instead, it resurrects it when control flow leaves the *next* unrelated catch block that happens to execute.
